thriftbp/client_pool.go

```diff
@@ -106,139 +113,69 @@ type ClientPool interface {
-func SingleAddressGenerator(addr string) AddressGenerator {
-	return func() (string, error) {
-		return addr, nil
-	}
-}
-
-// NewTTLClientFactory returns a ClientFactory that creates TTLClients using the
-// given TClientFactory.
-//
-// Services should generally not have to use NewTTLClientFactory directly,
-// instead you should use NewBaseplateClientPool which uses the default
-// ClientFactory for a typical Baseplate Thrift Client.
-func NewTTLClientFactory(ttl time.Duration) ClientFactory {
-	return func(clientFact TClientFactory, trans thrift.TTransport, protoFactory thrift.TProtocolFactory) Client {
-		return NewTTLClient(trans, clientFact(trans, protoFactory), ttl)
-	}
+	return newClientPool(cfg, genAddr, protoFactory, middlewares...)
 }
 
-// StandardTClientFactory returns a standard thrift.TClient using the given
-// thrift.TTransport and thrift.TProtocolFactory
-//
-// Services should generally not have to use StandardTClientFactory
-// directly, instead you should use NewBaseplateClientPool which uses the
-// default TClientFactory for a typical Baseplate Thrift Client.
-func StandardTClientFactory(trans thrift.TTransport, protoFactory thrift.TProtocolFactory) thrift.TClient {
-	return thrift.NewTStandardClient(
-		protoFactory.GetProtocol(trans),
-		protoFactory.GetProtocol(trans),
-	)
-}
-
-// NewWrappedTClientFactory returns a TClientFactory that returns a standard
-// thrift.TClient wrapped with the given middlewares.
-//
-// Services should generally not have to use NewWrappedTClientFactory directly,
-// instead you should use NewBaseplateClientPool which uses the default
-// TClientFactory for a typical Baseplate Thrift Client.
-func NewWrappedTClientFactory(base TClientFactory, middlewares ...thrift.ClientMiddleware) TClientFactory {
-	return func(trans thrift.TTransport, protoFactory thrift.TProtocolFactory) thrift.TClient {
-		client := base(trans, protoFactory)
-		return thrift.WrapClient(client, middlewares...)
-	}
-}
-
-// convenience struct for passing around the different factories needed to
-// create a Client.
-type factories struct {
-	Client   ClientFactory
-	TClient  TClientFactory
-	Protocol thrift.TProtocolFactory
-}
-
-func newClientPool(cfg ClientPoolConfig, genAddr AddressGenerator, factories factories) (*clientPool, error) {
+func newClientPool(
+	cfg ClientPoolConfig,
+	genAddr AddressGenerator,
+	proto thrift.TProtocolFactory,
+	middlewares ...thrift.ClientMiddleware,
+) (*clientPool, error) {
 	labels := cfg.MetricsLabels.AsStatsdLabels()
 	pool, err := clientpool.NewChannelPool(
 		cfg.InitialConnections,
 		cfg.MaxConnections,
 		func() (clientpool.Client, error) {
-			return newClient(cfg.SocketTimeout, genAddr, factories)
+			return newClient(
+				cfg.SocketTimeout,
+				cfg.MaxConnectionAge,
+				genAddr,
+				proto,
+				middlewares...,
+			)
 		},
 	)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("thriftbp: error initializing thrift clientpool. %w", err)
```

Comment:
🔕 I don't feel strongly about it but I think this error (and the ones in `newClient`) makes less sense to wrap. The context ones are generic that can happen everywhere so wrapping makes it easier to figure out where do they come from, but here the underlying error comes directly from `clientpool` (either `ErrExhausted`, `ConfigError` or network error from the underlying openers). But I guess it won't hurt to wrap them.

Reply:
To clarify you can totally ignore that comment, I was just thinking out loud. I don't think there's any real downside of wrapping them.